File: PolyScript/frameworks/test/TestCompiler.fs
(*
 * Test F# Compiler Tool for PolyScript Framework
 * CRUD × Modes Architecture: Zero-boilerplate CLI development
 * 
 * Author: Mathew Burkitt, Dimension Technologies <mathew.burkitt@ditech.ai>
 *)

open System
open System.Collections.Generic
open PolyScript.Framework

[<Rebadge("compile", "create+live")>]
[<Rebadge("dry-compile", "create+simulate")>]
[<Rebadge("status", "read+live")>]
[<Rebadge("clean", "delete+live")>]
type TestCompilerTool() =
    interface IPolyScriptTool with
        member _.Description = "Test F# compiler tool demonstrating CRUD × Modes"
        
        member _.Create resource options (context: PolyScriptContext) =
            let resourceName = resource |> Option.defaultValue "main.fs"
            context.Log(sprintf "Compiling %s..." resourceName)
            
            let outputFile = 
                match options.TryGetValue("output") with
                | (true, value) -> value :?> string
                | _ -> resourceName.Replace(".fs", ".exe")
            
            let optimize = 
                match options.TryGetValue("optimize") with
                | (true, value) -> value :?> bool
                | _ -> false
            
            dict [
                ("compiled", resourceName :> obj)
                ("output", outputFile :> obj)
                ("optimized", optimize :> obj)
                ("timestamp", DateTime.Now.ToString("O") :> obj)
            ] :> obj
        
        member _.Read resource options (context: PolyScriptContext) =
            context.Log("Checking compilation status...")
            
            let sourceFiles = 
                match resource with
                | Some r -> [r]
                | None -> ["Program.fs"; "Utils.fs"; "Config.fs"]
            
            dict [
                ("source_files", sourceFiles :> obj)
                ("compiled_files", ["Program.exe"; "Utils.dll"] :> obj)
                ("missing", ["Config.dll"] :> obj)
                ("last_build", DateTime.Now.ToString("O") :> obj)
            ] :> obj
        
        member _.Update resource options (context: PolyScriptContext) =
            let resourceName = resource |> Option.defaultValue "main.fs"
            context.Log(sprintf "Recompiling %s..." resourceName)
            
            let incremental = 
                match options.TryGetValue("incremental") with
                | (true, value) -> value :?> bool
                | _ -> false
            
            dict [
                ("recompiled", resourceName :> obj)
                ("reason", "source file changed" :> obj)
                ("incremental", incremental :> obj)
                ("timestamp", DateTime.Now.ToString("O") :> obj)
            ] :> obj
        
        member _.Delete resource options (context: PolyScriptContext) =
            let resourceName = resource |> Option.defaultValue "build artifacts"
            context.Log(sprintf "Cleaning %s..." resourceName)
            
            dict [
                ("cleaned", ["*.exe"; "*.dll"; "*.pdb"; "bin/"; "obj/"] :> obj)
                ("freed_space", "12.8 MB" :> obj)
                ("timestamp", DateTime.Now.ToString("O") :> obj)
            ] :> obj

[<EntryPoint>]
let main args =
    run<TestCompilerTool> args

File: PolyScript/frameworks/test/TestCompiler.scala
/*
 * Test Scala Compiler Tool for PolyScript Framework
 * CRUD × Modes Architecture: Zero-boilerplate CLI development
 * 
 * Author: Mathew Burkitt, Dimension Technologies <mathew.burkitt@ditech.ai>
 */

import polyscript.Framework._
import play.api.libs.json._
import scala.util.{Try, Success}
import java.time.Instant

object TestCompilerTool extends PolyScriptTool {
  def description: String = "Test Scala compiler tool demonstrating CRUD × Modes"
  
  def create(resource: Option[String], options: Map[String, Any], context: PolyScriptContext): Try[JsValue] = {
    context.log(s"Compiling ${resource.getOrElse("source")}...")
    
    Success(Json.obj(
      "compiled" -> resource.getOrElse("Main.scala"),
      "output" -> resource.map(_.replace(".scala", ".class")).getOrElse("Main.class"),
      "optimized" -> options.getOrElse("optimize", false).toString.toBoolean,
      "timestamp" -> Instant.now().toString
    ))
  }
  
  def read(resource: Option[String], options: Map[String, Any], context: PolyScriptContext): Try[JsValue] = {
    context.log("Checking compilation status...")
    
    Success(Json.obj(
      "source_files" -> Json.arr("Main.scala", "Utils.scala", "Config.scala"),
      "compiled_files" -> Json.arr("Main.class", "Utils.class"),
      "missing" -> Json.arr("Config.class"),
      "last_build" -> Instant.now().minusSeconds(7200).toString
    ))
  }
  
  def update(resource: Option[String], options: Map[String, Any], context: PolyScriptContext): Try[JsValue] = {
    context.log(s"Recompiling ${resource.getOrElse("source")}...")
    
    Success(Json.obj(
      "recompiled" -> resource.getOrElse("Main.scala"),
      "reason" -> "source file changed",
      "incremental" -> options.getOrElse("incremental", false).toString.toBoolean,
      "timestamp" -> Instant.now().toString
    ))
  }
  
  def delete(resource: Option[String], options: Map[String, Any], context: PolyScriptContext): Try[JsValue] = {
    context.log(s"Cleaning ${resource.getOrElse("build artifacts")}...")
    
    Success(Json.obj(
      "cleaned" -> Json.arr("*.class", "target/", ".bsp/"),
      "freed_space" -> "31.2 MB",
      "timestamp" -> Instant.now().toString
    ))
  }
}

object TestCompiler extends App {
  Framework.run(args, TestCompilerTool)
}